### PR TITLE
sql: split selectNode into a separate filterNode and optimize more.

### PR DESCRIPTION
**Pull request by knz**
tl;dr: query plans are simplified, become (probably) slightly faster.

The main achievement of this patch is the creation of a new
`filterNode` which takes over the filtering part of the WHERE
clause. The motivation for this is to support a subsequent commit
which will add filtering to arbitrary nodes but where rendering is not
needed.

What this new `filterNode` immediately enables is the opportunity to
elide the `selectNode` altogether when it is a no-op, namely when it
renders exactly its source columns without reordering or additional
operations. This optimization is also implemented here right away
because not doing it would make most of the EXPLAIN tests much more
verbose (due to a new filtering stage), whereas doing it makes most of
the tests *less* verbose than prior to this patch (because there were
many no-op render stages).

Another side-car of this patch is `selectTopNode` elision, since
`selectTopNode` is not really needed any more after plan expansion.

Since no-op `selectNode`s are now elided during expandPlan, the code
that was attempting to do the same in the distSQL physical planner can
be removed. This is a good thing, because I believe there was a bug in
there: it seems that a query of the form `SELECT v as k, k as v FROM
kv` was not handled properly, because the code was merely checking
that the column names and types of the selectNode and its source
matched, without checking that the referenced columns where in the
same order. The new selectNode elision optimization does check this
properly.

Two limitations remaining in this patch:

- the selectNode elision reveals new opportunities to optimize
  `scanNode`s, for example `SELECT * FROM (SELECT * FROM t) WHERE a >
  10` is optimized to `filterNode(a>10, scanNode(t))` which could
  become `scanNode(t, span = /10-/11)` but this remaining optimization
  is not performed yet. This is because of the traversal order, the
  scanNode has been optimized already by the time expandPlan decides
  to elide selectNode. A later patch could change the traversal
  algorithm to achieve this.

- the `selectNode` that results from the expansion of a view name
  reference into a sub-plan is not elided, even when it is a
  no-op. This is because CREATE VIEW needs to see, inside the
  `dataSourceInfo` attribute that it carries, that it was referencing
  a view to determine the view dependency graph. A later patch could
  make selectNode elision conditional and have CREATE VIEW set the
  flag that disables it.

Needed for #10633.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/12616)

### CODE REVIEW

**Comment by knz** [comment]
@RaduBerinde I am not very confident I am doing the right thing with the distsql physical planner. If you have any suggestions that would be great.

**Comment by RaduBerinde** [comment]
I didn't have time to look over the code in detail yet but I have a few questions:
 - why do you say that queries are now faster? Those that do both rendering and filtering now have data moving through more `Next()` calls.. Doing the filtering right after rendering is probably more efficient..
 - do we now add more renders to the select node as needed by the filter? (e.g. for `SELECT x FROM t WHERE y > 5`). And if so, shouldn't the `filterNode` remove those extra renders?
 - I haven't looked at the distsql planning stages but I want to point out that most processors support integrated filtering, it's important to use those as much as possible instead of adding extra stages just for filtering (since each stage involves a goroutine that moves data around).

Please also take a look at #12547, as there is some overlap (esp. the first commit). Maybe pull in that branch and see if you can work on top of it?

**Comment by knz** [comment]
>   * why do you say that queries are now faster? Those that do both
>     rendering and filtering now have data moving through more |Next()|
>     calls.. Doing the filtering right after rendering is probably more
>     efficient..

The selectTopNode is always elided so we always win one stage.

Then for most queries the filterNode is merged into the underlying
scanNode, then there's no difference, or the selectNode is actually
rendering the source columns (select * from ...) and is also elided,
then there's no difference, or both, in which case we have another one
stage fewer.

>   * do we now add more renders to the select node as needed by the
>     filter? (e.g. for |SELECT x FROM t WHERE y > 5|). And if so,
>     shouldn't the |filterNode| remove those extra renders?

No we don't add renders for WHERE expressions, because these expressions
are linked to the data source, not the renders (this is different from
GROUP BY / ORDER BY).

>   * I haven't looked at the distsql planning stages but I want to point
>     out that most processors support integrated filtering, it's
>     important to use those as much as possible instead of adding extra
>     stages just for filtering (since each stage involves a goroutine
>     that moves data around).

Well the new code should be simpler, so there's that.

> Please also take a look at #12547
> <https://github.com/cockroachdb/cockroach/pull/12547>, as there is some
> overlap (esp. the first commit). Maybe pull in that branch and see if
> you can work on top of it?

Will do

**Comment by knz** [comment]
I have rebased off your equivalent change in #12547. I now understand better what needed to be done. Let me know if there's anything else.

**Comment by RaduBerinde** [comment]
LGTM, cool stuff!

Please run `COCKROACH_DISTSQL_MODE=on COCKROACH_DISTSQL_LOG_PLAN=1 make test PKG=./pkg/sql TESTS=TestLogic` to make sure the distsql stuff still works (we don't yet have a test that runs automatically, sorry). There will be a failure for `/views`, ignore it (it's that issue we've been discussing).

Ok to merge together with my commit when ready.

---

Reviewed 1 of 2 files at r3, 34 of 34 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 162 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YK9zpo3R7xj4EtJe3:-K_YK9zpo3R7xj4EtJe4:b8knjk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/distsql_physical_planner.go#L162)):*
> ```Go
> 
> 	case *filterNode:
> 		if n.filter != nil {
> ```

Is it ever possible to have a `filerNode` with no filter? If no, remove this check..

---

*[pkg/sql/distsql_physical_planner.go, line 183 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YKr6Nn6ckHSdVZ6b3:-K_YKr6Nn6ckHSdVZ6b4:bw96qdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/distsql_physical_planner.go#L183)):*
> ```Go
> 		return dsp.CheckSupport(n.source.plan)
> 
> 	case *sortNode:
> ```

We aren't calling this on `n.sort` in the `selectTopNode` case above. We need to "OR" the `shouldDistribute` result between the source and the sort case like we did before

---

*[pkg/sql/distsql_physical_planner.go, line 1149 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YLGf8pkFhsffwSWwY:-K_YLGf8pkFhsffwSWwZ:b-vtkyp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/distsql_physical_planner.go#L1149)):*
> ```Go
> }
> 
> func (dsp *distSQLPlanner) createPlanForNode(
> ```

It seems like this fails if we have a `selectTopNode`..

---

*[pkg/sql/expand_plan.go, line 286 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YMSiDw68nWP9upck0:-K_YMSiDw68nWP9upck1:b-yf2qs7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/expand_plan.go#L286)):*
> ```Go
> 	maybeScanNode := s.source.plan
> 	var whereFilter parser.TypedExpr
> 	where, ok := maybeScanNode.(*filterNode)
> ```

Do we ever have a `scanNode` under a `filterNode`? That seems less than ideal, given that `scanNode` also has the ability to filter..

---

*[pkg/sql/expand_plan.go, line 333 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YMDljkSI3vKc49qlS:-K_YMDljkSI3vKc49qlT:bjb1n1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/expand_plan.go#L333)):*
> ```Go
> 	}
> 
> 	// Elide the render node if it renders its source as-is.
> ```

Nice! I also like the name "render node", seems like a much better name for what `selectNode` does.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-:-K_YLUSBT8YgR4G5rA-D:b8j3rxk)*

**Comment by knz** [comment]
I have ran the test and there was one last thing to fix (recursion in pre-expanded groupNode without selectTopNode). Other than that the distsql plans look fine.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 162 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YK9zpo3R7xj4EtJe3:-K_ZOCCcasjVTeIax48g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/distsql_physical_planner.go#L162)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Is it ever possible to have a `filerNode` with no filter? If no, remove this check..
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 183 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YKr6Nn6ckHSdVZ6b3:-K_ZRqQBF_IOgvKtjEcO:b-lf31d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/distsql_physical_planner.go#L183)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We aren't calling this on `n.sort` in the `selectTopNode` case above. We need to "OR" the `shouldDistribute` result between the source and the sort case like we did before
</blockquote></details>

I removed the selectTopNode case entirely, since it's never there any more when we get at this function.

---

*[pkg/sql/distsql_physical_planner.go, line 1149 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YLGf8pkFhsffwSWwY:-K_ZS-o3Ll9gIg---2fO:b-dnc28w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/distsql_physical_planner.go#L1149)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It seems like this fails if we have a `selectTopNode`..
</blockquote></details>

Does it? I don't see how.

---

*[pkg/sql/expand_plan.go, line 286 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YMSiDw68nWP9upck0:-K_ZEstZ3puuEl0ouF28:b-wrfy21) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/expand_plan.go#L286)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do we ever have a `scanNode` under a `filterNode`? That seems less than ideal, given that `scanNode` also has the ability to filter..
</blockquote></details>

This currently happens with EXPLAIN(NOOPTIMIZE,NOEXPAND).
Perhaps it won't if/when we simplify EXPLAIN but I didn't want to make too many changes at once.

---

*[pkg/sql/expand_plan.go, line 333 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YMDljkSI3vKc49qlS:-K_ZFCFGOX6Bd0DR0cxX:b-g6ulu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/expand_plan.go#L333)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Nice! I also like the name "render node", seems like a much better name for what `selectNode` does.
</blockquote></details>

Yeah I will also push another commit soon to rename the node entirely.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-:-K_ZS3RnzJGlzuxrRQtO:bu4r2od)*

**Comment by RaduBerinde** [comment]
*[pkg/sql/distsql_physical_planner.go, line 1149 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616#-K_YLGf8pkFhsffwSWwY:-K__uhb1-jeK2UQvUpwM:bybrcxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b7042be28f16ea0e23ae966fca5eb401bda7819/pkg/sql/distsql_physical_planner.go#L1149)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Does it? I don't see how.
</blockquote></details>

It would go to the default case which is a panic. But now it's all good since `CheckSupport` won't allow the query if there is a `selectTopNode`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12616)*

**Comment by knz** [comment]
Yeah I saw that too, that's why I simplified CheckSupport :)

**Comment by knz** [comment]
TFYR!
